Problem inserting valueBlobFile into bytea column postgres

## Environment

**Liquibase Version**:
Liquibase 4.14

**Liquibase Integration & Version**: 
CLI 

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**:
PostgreSQL 14.4

Via docker:
version: '3.8'

services:
  postgres:
    image: postgres:14.4
    ports:
      - 5432:5432
    volumes:
      - /d/postgres/exampledb:/var/lib/postgresql/data
    environment:
      - POSTGRES_USER=admin
      - POSTGRES_PASSWORD=4321
      - POSTGRES_DB=exampledb
      - PGDATA=/var/lib/postgresql/data/pgdata 


**Operating System Type & Version**:
Windows 10 

**Infrastructure Type/Provider**:
CLI

## Description
Following Changelog runs without problems with liquibase 4.11:

<?xml version="1.0" encoding="UTF-8"?>
<databaseChangeLog
	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
	xmlns:pro="http://www.liquibase.org/xml/ns/pro"
	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd
		http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd
		http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-latest.xsd">

	<changeSet id="di-1" author="someauthor">
		<createTable tableName="blobtest">
			<column name="id" type="int8">
				<constraints nullable="false" />
			</column>
			<column name="content" type="bytea" />
		</createTable>

		<addPrimaryKey columnNames="id"
			constraintName="blobtest_pk" tableName="blobtest" />

		<createSequence sequenceName="blobtest_sequence"
			incrementBy="50" startValue="100" />
	</changeSet>

	<changeSet id="di-2" author="someauthor">
		<insert tableName="blobtest">
			<column name="id" valueNumeric="1" />
			<column name="content" valueBlobFile="foo.txt" />
		</insert>
	</changeSet>

</databaseChangeLog>

foo.txt is textfile UTF-8 endoded content: foo

With 4.14 i get an Exception:
Caused by: org.postgresql.util.PSQLException: ERROR: column "content" is of type bytea but expression is of type bigint
  Hinweis: You will need to rewrite or cast the expression.

## Steps To Reproduce
Use the CLI with the changelog above against an postgres db.

## Actual Behavior
An exception in thrown with the cause stated above.

## Expected/Desired Behavior
A changelog with bytea as type and valueBlobFile creates a table with a column of type bytea and the

I have the same bug when trying to updaload data in my table.
The pb has been introduce by the 4.12.0 (and is still present with more recent version), it works fine with 4.11.0 as [jnclmns](https://github.com/jnclmns).


My context SpringBoot 2.7.3
Liquibase 4.12.0
DB posgresql

My table:
```
  - changeSet:
      id: 1 - translation
      changes:
        - createTable:
            tableName: TRANSLATION
            columns:
              - column:
                  name: ID
                  type: bigint
                  constraints:
                    primaryKey: true
                    nullable: false
              - column:
                  name: UID
                  type: varchar(40)
              - column:
                  name: PDF_TMP
                  type: bytea
                  
  - changeSet:
      id: 2 - Add Translation data
      context: sample_data
      changes:
        - loadData:
            encoding: UTF-8
            file: db/changelog/data/sample_data/2_translation.csv
            separator: ","
            tableName: TRANSLATION
            usePreparedStatements: true
            columns:
              - column:
                  name: PDF_TMP
                  type: BLOB
```

With this csv file: `2_translation.csv`
```
ID,PDF_TMP,UID
-10001,data/sample_data/100_de.pdf,AAAAAAAAAAA
-10002,data/sample_data/100_en.pdf,BBBBBBBBBBB
-10003,data/sample_data/100_fr.pdf,CCCCCCCCCCC
```

And I have the same error :
```
Caused by: org.postgresql.util.PSQLException: ERROR: column "pdf_tmp" is of type bytea but expression is of type bigint
  Hint: You will need to rewrite or cast the expression.
  Position: 139
```

Same issue found for clob file too. <column name="colNmae" valueClobFile="clob/clobfile.txt"/> gives similar error:

Caused by: org.postgresql.util.PSQLException: ERROR: column "columnPK" is of type bigint but expression is of type character varying
  Hint: You will need to rewrite or cast the expression.
  Position: 264

--- a/liquibase-core/src/main/java/liquibase/statement/ExecutablePreparedStatementBase.java
+++ b/liquibase-core/src/main/java/liquibase/statement/ExecutablePreparedStatementBase.java
@@ -15,8 +15,10 @@ import liquibase.exception.LiquibaseException;
 import liquibase.exception.UnexpectedLiquibaseException;
 import liquibase.listener.SqlListener;
 import liquibase.logging.Logger;
-import liquibase.resource.InputStreamList;
 import liquibase.resource.ResourceAccessor;
+import liquibase.snapshot.SnapshotGeneratorFactory;
+import liquibase.structure.core.Column;
+import liquibase.structure.core.Table;
 import liquibase.util.FilenameUtil;
 import liquibase.util.JdbcUtil;
 import liquibase.util.StreamUtil;
@@ -56,6 +58,8 @@ public abstract class ExecutablePreparedStatementBase implements ExecutablePrepa
 
     //Cache the executeWithFlags method to avoid reflection overhead
     private static Method executeWithFlagsMethod;
+
+    private Map<String, Object> snapshotScratchPad = new HashMap<>();
 
     protected ExecutablePreparedStatementBase(Database database, String catalogName, String schemaName, String
             tableName, List<? extends ColumnConfig> columns, ChangeSet changeSet, ResourceAccessor resourceAccessor) {
@@ -272,6 +276,25 @@ public abstract class ExecutablePreparedStatementBase implements ExecutablePrepa
             LOG.fine("value is blob = " + col.getValueBlobFile());
             try {
                 LOBContent<InputStream> lob = toBinaryStream(col.getValueBlobFile());
+
+                if (database instanceof PostgresDatabase) {
+                    String snapshotKeyName = String.format("%s-%s-%s-%s", getCatalogName(), getSchemaName(), getTableName(), col.getName());
+                    Column snapshot = (Column) this.getScratchData(snapshotKeyName);
+                    if (snapshot == null) {
+                        snapshot = SnapshotGeneratorFactory.getInstance().createSnapshot(
+                                new Column(Table.class, getCatalogName(), getSchemaName(), getTableName(), col.getName()), database);
+                        this.setScratchData(snapshotKeyName, snapshot);
+                    }
+
+                    if (snapshot.getType().getTypeName().equalsIgnoreCase("bytea")) {
+                        if (lob.length <= Integer.MAX_VALUE) {
+                            stmt.setBinaryStream(i, lob.content, (int) lob.length);
+                        } else {
+                            stmt.setBinaryStream(i, lob.content, lob.length);
+                        }
+                        return;
+                    }
+                }
                 if (lob.length <= Integer.MAX_VALUE) {
                     stmt.setBlob(i, lob.content, (int) lob.length);
                 } else {
@@ -460,6 +483,14 @@ public abstract class ExecutablePreparedStatementBase implements ExecutablePrepa
 
     public ResourceAccessor getResourceAccessor() {
         return resourceAccessor;
+    }
+
+    private Object getScratchData(String key) {
+        return snapshotScratchPad.get(key);
+    }
+
+    private Object setScratchData(String key, Object data) {
+        return snapshotScratchPad.put(key, data);
     }
 
     protected long getContentLength(InputStream in) throws IOException {

--- a/liquibase-integration-tests/src/test/java/liquibase/dbtest/pgsql/PostgreSQLIntegrationTest.java
+++ b/liquibase-integration-tests/src/test/java/liquibase/dbtest/pgsql/PostgreSQLIntegrationTest.java
@@ -22,16 +22,19 @@ import org.junit.Assert;
 import org.junit.Test;
 
 import java.util.List;
+import java.util.Map;
 
 import static org.junit.Assume.assumeNotNull;
 
 public class PostgreSQLIntegrationTest extends AbstractIntegrationTest {
 
-    String dependenciesChangeLog = null;
+    private String dependenciesChangeLog = null;
+    private String blobChangeLog = null;
 
     public PostgreSQLIntegrationTest() throws Exception {
         super("pgsql", DatabaseFactory.getInstance().getDatabase("postgresql"));
         dependenciesChangeLog = "changelogs/pgsql/complete/testFkPkDependencies.xml";
+        blobChangeLog = "changelogs/pgsql/complete/testBlob.changelog.xml";
     }
 
     /**
@@ -67,6 +70,25 @@ public class PostgreSQLIntegrationTest extends AbstractIntegrationTest {
                           .findFirst()
                           .orElse(null);
             Assert.assertNull(addPrimaryKeyChangeSet);
+        } catch (ValidationFailedException e) {
+            e.printDescriptiveError(System.out);
+            throw e;
+        }
+    }
+
+    @Test
+    public void testBlobTypesChangeLog() throws Exception {
+        assumeNotNull(this.getDatabase());
+        Liquibase liquibase = createLiquibase(this.blobChangeLog);
+        clearDatabase();
+        try {
+            liquibase.update();
+            List<Map<String, ?>>  data = Scope.getCurrentScope().getSingleton(ExecutorService.class).getExecutor("jdbc", getDatabase())
+                    .queryForList(
+                            new RawSqlStatement("SELECT pg_column_size(content_bytea) as BYTEASIZE, pg_column_size(lo_get(content_oid)) as OIDSIZE FROM  public.blobtest"));
+            Assert.assertNotNull(data.get(0));
+            Assert.assertTrue(((Integer)data.get(0).get("BYTEASIZE")) > 0);
+            Assert.assertEquals(data.get(0).get("BYTEASIZE"), data.get(0).get("OIDSIZE"));
         } catch (ValidationFailedException e) {
             e.printDescriptiveError(System.out);
             throw e;

--- a/liquibase-integration-tests/src/test/resources/changelogs/pgsql/complete/testBlob.changelog.xml
+++ b/liquibase-integration-tests/src/test/resources/changelogs/pgsql/complete/testBlob.changelog.xml
@@ -1,0 +1,22 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-4.1.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd">
+
+    <changeSet id="di-1" author="someauthor">
+        <createTable tableName="blobtest">
+            <column name="id" type="int8">
+                <constraints nullable="false"/>
+            </column>
+            <column name="content_bytea" type="bytea"/>
+            <column name="content_oid" type="oid"/>
+        </createTable>
+    </changeSet>
+
+    <changeSet id="di-2" author="someauthor">
+        <insert tableName="blobtest">
+            <column name="id" valueNumeric="1"/>
+            <column name="content_bytea" valueBlobFile="datafiles/enum-data.csv" />
+            <column name="content_oid" valueBlobFile="datafiles/enum-data.csv"/>
+        </insert>
+    </changeSet>
+
+</databaseChangeLog>

